Ported to 2.80 syntax

wizardgsz opened the pull request:
Please review my contrib, if you have time obviously :D
Kind regards

wizardgsz committed:
Ported to 2.80 syntax

RichysHub commented on io_scene_vox.py:
```diff
@@ -203,53 +220,147 @@ def import_vox(path, *, voxel_spacing=1, voxel_size=1,
+#     Register - Unregister - MAIN
+#==================================================================================================
+#
+__classes__ = (
```
I don't like creating new dunder names, please refactor

wizardgsz replied:
I got it.

RichysHub commented on io_scene_vox.py:
```diff
@@ -61,38 +71,43 @@
+    gamma_value: FloatProperty(name="Gamma Correction Value", default=2.2, min=0)
 
-    use_shadeless = BoolProperty(name="Use Shadeless Materials", default=False)
+    use_shadeless: BoolProperty(name="Use Shadeless Materials", default=False)
```
Unsure why these assignments were changed.
To me, these read as type hint formatting.

Please remove these changes or provide source for this format.

wizardgsz replied:
It is the new syntax for 2.80+

wizardgsz replied:
See also: https://b3d.interplanety.org/en/porting-add-on-from-blender-2-7-to-blender-2-8/

RichysHub replied:
Ahh, very good. Found the supporting literature in the docs, https://docs.blender.org/api/current/bpy.types.Operator.html

Did not know about that change, new syntax definitely correct here 👍

RichysHub commented on io_scene_vox.py:
```diff
@@ -203,53 +220,147 @@ def import_vox(path, *, voxel_spacing=1, voxel_size=1,
+            material.diffuse_color = [pow(col / 255, gamma_value) for col in palette_entry[:4]]
+            #material.diffuse_intensity = 1.0
+            #material.alpha = palette_entry[3]
+            #material.use_shadeless = use_shadeless
```
commented code is dead code, cleanup these please

RichysHub commented on io_scene_vox.py:
```diff
@@ -203,53 +220,147 @@ def import_vox(path, *, voxel_spacing=1, voxel_size=1,
+#     http://web.purplefrog.com/~thoth/blender/python-cookbook/change-diffuse-to-emission-node.html 
+#==================================================================================================
+#
+import bpy
```
No need to reimport

RichysHub commented on io_scene_vox.py:
```diff
@@ -203,53 +220,147 @@ def import_vox(path, *, voxel_spacing=1, voxel_size=1,
 def register():
-    bpy.utils.register_module(__name__)
-    bpy.types.INFO_MT_file_import.append(menu_func_import)
+    from bpy.utils import register_class
```
import statements at start of file, please

wizardgsz replied:
menu_func_import, register, unregister and the "main" calling register()?

Should not be at the end of file as they need other classes implemented? Excuse me but I never programmed in Python before.

RichysHub replied:
Apologies for lack of clarity.

`from bpy.utils import register_class`  is just importing this `register_class` function for us to use here.
Generally I prefer all the import statements to be collected at the top of the file.

Although, in this case, we could also replace the line `register_class(cls)` with `bpy.utils.register_class(cls)` and skip the import entirely (as we import the base `bpy` right at the start)

RichysHub commented on io_scene_vox.py:
```diff
@@ -203,53 +220,147 @@ def import_vox(path, *, voxel_spacing=1, voxel_size=1,
+    # Test call
+    #bpy.ops.import_scene.vox('INVOKE_DEFAULT')
+    #bpy.types.TOPBAR_MT_file_import.remove(menu_func_import)
+
```
Commented test code, please remove

RichysHub commented on io_scene_vox.py:
```diff
@@ -203,53 +220,147 @@ def import_vox(path, *, voxel_spacing=1, voxel_size=1,
+        mat.node_tree.nodes.remove(node)
+
+
+def replace_on_selected_objects():
```
This is never called, so can be removed for clarity

RichysHub commented on io_scene_vox.py:
```diff
@@ -203,53 +220,147 @@ def import_vox(path, *, voxel_spacing=1, voxel_size=1,
+    for mat in mats:
+        material_diffuse_to_emission(mat)
+
+def replace_in_all_materials():
```
Also never called, presumably came wholesale with the shadeless solution

wizardgsz replied:
Yeah, exactly. I just imported the whole code and functions.

RichysHub reviewed (changes requested):
Overall, thanks for taking the time to update, and submit here for others to get use from. 

I've notated some more specific areas that I would like reworked. Perhaps a little shuffling and tidying blank lines here and there is warranted too.

Also commit included ".vs/MagicaVoxel-VOX-importer/v16/.suo" and ".vs/slnx.sqlite" binary files, please remove from pull, and update .gitignore

wizardgsz committed:
.

wizardgsz committed:
.

wizardgsz committed:
Delete slnx.sqlite

wizardgsz committed:
Delete .suo

wizardgsz committed:
Reviewed code

wizardgsz committed:
Reviewed code

RichysHub commented on io_scene_vox.py:
```diff
@@ -203,52 +223,119 @@ def import_vox(path, *, voxel_spacing=1, voxel_size=1,
         copy = base_voxel.copy()
         copy.data = base_voxel.data.copy()
         copy.location = [float(coord) * voxel_spacing for coord in voxel[:3]]
+        #copy.parent = base_voxel
```
```suggestion
```

wizardgsz replied:
It was just a test to "hierarchical link" the base object you "clone" for each voxel. 
I think that base_voxel object should be deleted anyhow at the end of the import.

RichysHub replied:
I have had some thoughts about having a 'base model' be created, outside of the model, that all the voxels are cloned instances of. 

But that's beyond the scope of the port.

Currently it's commented out code, with no need to stick around

RichysHub commented on io_scene_vox.py:
```diff
@@ -203,52 +223,119 @@ def import_vox(path, *, voxel_spacing=1, voxel_size=1,
+        else:
+            connected_sockets_out.append(None)
+
+    #print( defaults_in )
```
```suggestion
```

RichysHub commented on io_scene_vox.py:
```diff
@@ -175,6 +194,7 @@ def import_vox(path, *, voxel_spacing=1, voxel_size=1,
                 binary = bin(prop_bits)
                 # Need to read property values, but this gets fiddly
                 # TODO: finish implementation
+                vox.seek(_offset + 12 + s_self, 0)
```
Can we discuss this a little more closely.

I don't see where the value of _offset would be coming from.

Looking at the old .vox spec, I do see that there is a "normalized property value" at the end of this chunk. I don't, however see where the N in "4 x N". I presume that's what you're combating here, and using the known size of self to handle it?

Elsewhere I have used a `vox.read(12)` to pass over 12 bytes, rather than using seek, would be good to have consistency one way or the other. seek's second argument, which you've set to 0 here could be 1, which would seek relative to current position.

Perhaps this line could read
```suggestion
                # We have read 16 bytes of this chunk so far, ignoring remainder
                vox.read(s_self - 16)
```

RichysHub replied:
I see you made changes to address this in [this commit](https://github.com/RichysHub/MagicaVoxel-VOX-importer/pull/8/commits/662f55ab9a3459ccb74d99b3efe4b3c489db4711), but I'm not totally happy with that solution.

Recreating an `_offset` every chunk, when we only use it in the MATT seems fruitless to me, especially when we can just read ahead the remaining bytes in s_self.

wizardgsz replied:
Ouch, you are right: here I decided to skip the MATT chunk at all, as you implentation was still "in progress", but I forgot to copy the ftell function call before... 

In pseudo code:

```
while(true)
{
    _offset=ftell()
    ...
    switch(chunk type)
    {
      case PACK:
        ...parse it...
        break
      case MATT:
        fseek(_offset + 12 + chunk size, 0)
        break
      case ...:
      ...
    }
}
```

I tested different files, for example [red_booth_transparency.vox](https://github.com/kluchek/vox-models/blob/master/%23redbooth/%23red_booth_transparency.vox)



```
Reading at dec: 20 hex: 14
Parsing chunk 'PACK':
Reading at dec: 36 hex: 24
Parsing chunk 'SIZE':
Reading at dec: 60 hex: 3C
Parsing chunk 'XYZI':
Reading at dec: 12748 hex: 31CC
Parsing chunk 'RGBA':
Reading at dec: 13784 hex: 35D8
Parsing chunk 'MATT':
Reading at dec: 13832 hex: 3608
Parsing chunk 'MATT':
Reading at dec: 13872 hex: 3630
Parsing chunk 'MATT':
Reading at dec: 13916 hex: 365C
Parsing chunk 'MATT':
Reading at dec: 13964 hex: 368C
Loaded voxels: 3167
Linked voxels: 3000
```

RichysHub reviewed (changes requested):
I would strongly also suggest you add the entire .vs directory to the .gitignore file, to save accidentally adding it with each code commit

wizardgsz committed:
Skip MATT chunk

wizardgsz committed:
Update io_scene_vox.py

Co-Authored-By: RichysHub <richy_s@hotmail.co.uk>

wizardgsz committed:
Update io_scene_vox.py

Co-Authored-By: RichysHub <richy_s@hotmail.co.uk>

wizardgsz committed:
.

wizardgsz committed:
.

wizardgsz committed:
Update io_scene_vox.py

It works:

Reading at dec: 20 hex: 14
Parsing chunk 'PACK':
Reading at dec: 36 hex: 24
Parsing chunk 'SIZE':
Reading at dec: 60 hex: 3C
Parsing chunk 'XYZI':
Reading at dec: 12748 hex: 31CC
Parsing chunk 'RGBA':
Reading at dec: 13784 hex: 35D8
Parsing chunk 'MATT':
Reading at dec: 13832 hex: 3608
Parsing chunk 'MATT':
Reading at dec: 13872 hex: 3630
Parsing chunk 'MATT':
Reading at dec: 13916 hex: 365C
Parsing chunk 'MATT':
Reading at dec: 13964 hex: 368C

Co-Authored-By: RichysHub <richy_s@hotmail.co.uk>

wizardgsz commented:
Instead of .gitignore, I'm using /git/info/exclude now

wizardgsz committed:
Removed vox.tell()

RichysHub commented:
I'm going to go over this a couple more times, just to check I'm happy with everything.
Also going to manually test the imports (because this repository is lacking proper testing).

wizardgsz commented:
Maybe you already have enough "test cases", I downloaded sample data from:

https://github.com/mikelovesrobots/mmmm
https://github.com/ephtracy/voxel-model
https://github.com/marcosecchi/naba-2016-workshop-assets
https://github.com/kluchek/vox-models

Btw, I found this interesting parser also (unfortunately it does not handle new chunks):
[MagicaVoxel File format spec for Kaitai Struct](https://formats.kaitai.io/magicavoxel_vox/index.html)

RichysHub commented:
Oh, nice find on that parser (though it's not a system I'm familiar with)

I developed this based on ephtracy's [original spec](https://github.com/ephtracy/voxel-model/blob/master/MagicaVoxel-file-format-vox.txt), and have had a rewrite on the horizon to [the new spec](https://github.com/ephtracy/voxel-model/blob/master/MagicaVoxel-file-format-vox-extension.txt).

RichysHub commented on io_scene_vox.py:
```diff
@@ -203,21 +224,23 @@ def import_vox(path, *, voxel_spacing=1, voxel_size=1,
     location = [float(coord) * voxel_spacing for coord in voxel[:3]]
     # Using primitive_cube_add once here, to give us a template cube
-    bpy.ops.mesh.primitive_cube_add(radius=0.5 * voxel_size, location=location)
+    bpy.ops.mesh.primitive_cube_add(size=0.5 * voxel_size, location=location)
```
![image](https://user-images.githubusercontent.com/20910418/65823862-28bc3a80-e256-11e9-8039-d5f73b3c6e50.png)

Left is MagicaVoxel, right is the blender import of the same file. When the cube primitive used to take a radius, 0.5 was correct to get unit side-length voxels. That looks to have changed with the 2.80 API.

```suggestion
    bpy.ops.mesh.primitive_cube_add(size=voxel_size, location=location)
```

wizardgsz replied:
Yeah, in fact I am using "Voxel Size: 2.0" in the Importer options to load files. I though it was your personal choice to draw cubes this way :)

RichysHub commented:
![image](https://user-images.githubusercontent.com/20910418/65823905-1c84ad00-e257-11e9-98ec-c853f1cf3ee6.png)

I've not used Blender 2.80 at all, but this feels less than ideal as an end result for the scene collection.

![image](https://user-images.githubusercontent.com/20910418/65823914-5bb2fe00-e257-11e9-8b94-88ee4de18b31.png)

This (manually edited) result feels more "correct" to me? Not sure how feasible this would be within the current framework.

wizardgsz commented:
I think we could add every voxel in a new "collection" (the collection is a new structure in 2.80), exactly as you planned in the above screenshot (and remove the "base" cloned cube).
Official SVG importer, for example, works this way. You are right.

Something like this:

# create new collection
    name = os.path.basename(path)
    collection = bpy.data.collections.new(name)
    bpy.context.scene.collection.children.link(collection)

...and modify the later "link" to scene collection

    for object_ in to_link:
        collection.objects.link(object_)

wizardgsz committed:
Update io_scene_vox.py

Co-Authored-By: RichysHub <richardjspencer93@gmail.com>

wizardgsz committed:
Link imported VOX to a new "collection"

wizardgsz commented:
Moreover, I'd like to "collapse" the newly created collection, but it seems there is no API for it yet :(

There is a workaround but I do not like to use workarounds usually:
https://devtalk.blender.org/t/view-layer-api-access-wishlist-collection-expand-set/5517/3
https://blenderartists.org/t/question-regarding-expanding-collapsing-collection-in-outliner-in-2-8/1175242

wizardgsz commented:
Should we delete "base_voxel" at the end?

```
def delete_object(obj):
    # Deselect all
    bpy.ops.object.select_all(action='DESELECT')
    obj.select_set(True)
    # Call the operator only once
    bpy.ops.object.delete()
```

Instead of (de)selecting the object(s) beforehand, you can also pass a custom context (a list of objects) to avoid changing the selection within the viewport:

```
# Delete the template cube
bpy.ops.object.delete({"selected_objects": [base_voxel]})
```

wizardgsz committed:
Delete the template cube

wizardgsz commented:
About "default palette": please double check the unpack big/little endian mode, I think it should be corrected in

struct.unpack('<4B', struct.pack(**'<I'**, DEFAULT_PALETTE[col]))

I verified it loading this VOX without custom paletter i.e. RGBA chunk: [char_cat.vox](https://github.com/ephtracy/voxel-model/raw/master/vox/character/chr_cat.vox)

```
    if not palette:  # no palette provided, use default
        for col in range(256):
            palette.update({col + 1: struct.unpack('<4B', struct.pack('<I', DEFAULT_PALETTE[col]))})
```

RichysHub commented:
Okay, have taken some time to look over the changes.

- Very happy with the collection organisation, definitely an improvement. 👍
- I agree on not using a workaround for something that's non-critical (collapsing group) ✔
- Deleting of the initial cube, though annoying, might be the simplest way to end clean 🆗
- Current implementation doesn't leave a direct copy of the initial voxel (the one we clone). Probably best to just not peel the first voxel data at all, and create the cube at origin. ❗❗

![image](https://user-images.githubusercontent.com/20910418/65840343-d5a6be00-e30f-11e9-97cd-8b26f6e53144.png)

- With regards to color, this came up in issue #5, and was inverted in commit [9e79292](https://github.com/RichysHub/MagicaVoxel-VOX-importer/commit/9e792920fc9c91ec0d07a8226c96ce5a9ea17e52). I would suggest some thorough investigation here, because I am unsure. ❔

- Small research into point above made me realise ` material.diffuse_color = [pow(col / 255, gamma_value) for col in palette_entry[:4]]` is gamma correcting the alpha channel, which I believe is incorrect. ❗

RichysHub commented on io_scene_vox.py:
```diff
@@ -203,21 +224,29 @@ def import_vox(path, *, voxel_spacing=1, voxel_size=1,
+
     # Using primitive_cube_add once here, to give us a template cube
-    bpy.ops.mesh.primitive_cube_add(radius=0.5 * voxel_size, location=location)
+    bpy.ops.mesh.primitive_cube_add(size=voxel_size, location=location)
```
Just create the cube at 0,0,0. It's going to be deleted now anyway
```suggestion
    bpy.ops.mesh.primitive_cube_add(size=voxel_size)
```

RichysHub commented on io_scene_vox.py:
```diff
@@ -203,21 +224,29 @@ def import_vox(path, *, voxel_spacing=1, voxel_size=1,
             mat_palette.update({index: material})
 
     # peel first voxel information
     voxel, *voxels = voxels
```
Don't peel first voxel, this was only useful when we cared about the first cube
```suggestion
```

RichysHub commented on io_scene_vox.py:
```diff
@@ -203,21 +224,29 @@ def import_vox(path, *, voxel_spacing=1, voxel_size=1,
+                material_diffuse_to_emission(material)
             mat_palette.update({index: material})
 
     # peel first voxel information
```
Remove dead comment ⚰
```suggestion
```

RichysHub commented on io_scene_vox.py:
```diff
@@ -203,21 +224,29 @@ def import_vox(path, *, voxel_spacing=1, voxel_size=1,
 
     # peel first voxel information
     voxel, *voxels = voxels
     location = [float(coord) * voxel_spacing for coord in voxel[:3]]
```
No longer caring about first voxel as anything special
```suggestion
```

RichysHub commented on io_scene_vox.py:
```diff
@@ -203,21 +224,29 @@ def import_vox(path, *, voxel_spacing=1, voxel_size=1,
-    bpy.ops.mesh.primitive_cube_add(radius=0.5 * voxel_size, location=location)
+    bpy.ops.mesh.primitive_cube_add(size=voxel_size, location=location)
     base_voxel = bpy.context.object
     if use_palette:
```
```suggestion
```

RichysHub commented on io_scene_vox.py:
```diff
@@ -203,21 +224,29 @@ def import_vox(path, *, voxel_spacing=1, voxel_size=1,
+    bpy.ops.mesh.primitive_cube_add(size=voxel_size, location=location)
     base_voxel = bpy.context.object
     if use_palette:
         base_voxel.active_material = mat_palette[voxel[3]]
```
```suggestion
```

RichysHub reviewed (commented):
Removing the special first voxel, making it instead a generic cube at 0,0,0.
This fixes issue with missing voxel in end result

wizardgsz committed:
Update io_scene_vox.py

Co-Authored-By: RichysHub <richardjspencer93@gmail.com>

wizardgsz committed:
Update io_scene_vox.py

Co-Authored-By: RichysHub <richardjspencer93@gmail.com>

wizardgsz committed:
Update io_scene_vox.py

Co-Authored-By: RichysHub <richardjspencer93@gmail.com>

wizardgsz committed:
Update io_scene_vox.py

Co-Authored-By: RichysHub <richardjspencer93@gmail.com>

wizardgsz committed:
Update io_scene_vox.py

Co-Authored-By: RichysHub <richardjspencer93@gmail.com>

wizardgsz committed:
Update io_scene_vox.py

Co-Authored-By: RichysHub <richardjspencer93@gmail.com>

wizardgsz commented:
Any other open point? Next, I'd like to work on material properties in **MATT** chunk...

wizardgsz commented:
> Removing the special first voxel, making it instead a generic cube at 0,0,0.
> This fixes issue with missing voxel in end result

I think we have to modify the "bounds" properties `start_voxel` and `end_voxel`. Are they 1-based in your implementation?

RichysHub commented on io_scene_vox.py:
```diff
@@ -203,20 +224,23 @@ def import_vox(path, *, voxel_spacing=1, voxel_size=1,
-            material.diffuse_intensity = 1.0
-            material.alpha = palette_entry[3]
-            material.use_shadeless = use_shadeless
+            material.diffuse_color = [pow(col / 255, gamma_value) for col in palette_entry[:4]]
```
Alpha channel is still being gamma corrected here, which it should not be.

RichysHub commented:
```
    if use_bounds:
        # clamp end_voxel to size of model
        end = min([end_voxel, len(voxels)])
        voxels = voxels[start_voxel:end]
```
Voxel bounds are definitely implemented as zero based, but then the slider in the UI defaults to 1, and is min of 1. 

Well spotted! 👍

Bounds were added to have a chance of loading big models, so weren't implemented with much care. I have no strong feelings either way, but yes, we do need to unify the two. Also a label or something in the UI might be good to inform.

wizardgsz committed:
Adjust voxels bounds

wizardgsz committed:
Do not "gamma correct" palette alpha values

RichysHub commented:
Okay, I don't think there's anything outstanding that needs doing.

I'll wait until I have your okay to do so, but I think this is ready to be squashed together and merged in. 🎉

wizardgsz commented:
Nice, thanks! I think we are ready to close the pull. I'll open a new request for **MATT/MATL** and new chunks if you agree.

RichysHub commented:
Hmm, I wonder if there's something smarter we could do here.
There's a lot of commits in this pull (25) which would make it more than 2/3 of the total repo commits. I don't want to just merge them all in, and leave a messy git history.
So I was about to squash this whole thing down into a single commit, and merge it in.

However, that leaves anyone using 2.74 having to use the older version.
That's fine, and I intended just to tag the relevant point, maybe make it a release on GitHub to make it super clear for anyone still working <2.80.

But, in this pull request you've also identified and fixed a couple bugs (voxel bounds, `MATT` chunk skipping, and palette).
Ideally those would be fixed in the 2.74 leave-off version.

I don't know how much control you have with regards to rearranging commits, and I'm not a git guru myself.

If we could get this pull down to just:
- fixing bugs in 2.74 (either 1 commit, or a couple)
- porting to 2.8

Then we could merge all the commits separate, have a 2.74 point with fixed bugs, and have a clean history.

I'm going to do some reading, see if I can cherry pick out the bugfix commits and patch them on first.

wizardgsz commented:
There is an option to check for Blender version using `bpy.app.version`:
```
if (2, 74, 0) < bpy.app.version:
    print("Here it is the old code")
```

But I think the code will be less "intuitive" to read and maybe harder to maintain

I agree: you can start "fixing" the few minor bugs (the palette should be ok in your old implementation for 2.74, I added a bug gamma-correcting apha values instead). 

I am not a "Git" expert unfortunately: it is the first time I use it (I am used to the old SourceSafe or later TFS by Microsoft).

RichysHub commented:
Alright, I'll get reading.

Once we start adding new features to the 2.80 importer, I'll put some brain power towards how we could have a single script work for both, without being a huge ugly mess.

RichysHub committed:
Fixed default palette implementation. This was changed in 9e79292, thinking that it was otherwise broken. Conversation with wizardgsz in issue #5 has revealed that this change was in error. This commit returns it to correct.

RichysHub commented:
Okay, I managed to move things around to get these changes merged into master.
Separated out your bugfixes, so hopefully Blender <2.80 can enjoy them.
Have tagged 2 releases, for the relevant Blender versions.

I have had to rebase some things, so if you have built upon the code in this pull request, you will have to rebase onto the master branch. I can help you with that in the relevant issue.

Thank you for your hard work on getting this pull through 🎉🎊
